elasticai_api/common/data_shard_service.py

```diff
@@ -134,3 +137,51 @@ def fetch_shard(self):
             return None
 
         return task.shard
+
+
+class RecordIndexService(DataShardService):
+    def __init__(
+        self,
+        master_client,
+        batch_size,
+        num_epochs=None,
+        dataset_size=None,
+        task_type=elasticai_api_pb2.TRAINING,
+        shuffle=False,
+    ):
+        super(RecordIndexService, self).__init__(
+            master_client, batch_size, num_epochs, dataset_size, task_type
+        )
+        self._shuffle = shuffle
+        self._shard_queue = SimpleQueue()
+        threading.Thread(
+            target=self._get_shard_indices,
+            name="check_timeout_tasks",
+            daemon=True,
+        ).start()
+
+    def _get_shard_indices(self):
+        while True:
+            if self._shard_queue.empty():
```

Comment:
Don't need to wait for empty, if the queue size is small, such as   `< batchsize`, get next shard.

```
 if self._shard_queue.qsize() < batchsize: